Introduce NoVirtualDestructorBase and use it in various WTF classes

rniwa opened the pull request:
#### 66366e42f025467527b81dd44c77519dabc86152
<pre>
Introduce NoVirtualDestructorBase and use it in various WTF classes
<a href="https://bugs.webkit.org/show_bug.cgi?id=290651">https://bugs.webkit.org/show_bug.cgi?id=290651</a>

Reviewed by Chris Dumez.

Introduce NoVirtualDestructorBase and replace the use of
SUPPRESS_REFCOUNTED_WITHOUT_VIRTUAL_DESTRUCTOR with it in various WTF classes.

Suppress the unsafe cast warning in SymbolImpl::StaticSymbolImpl now that
this class is checked by the static analyzer.

* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/CMakeLists.txt:
* Source/WTF/wtf/Compiler.h:
* Source/WTF/wtf/JSONValues.h:
* Source/WTF/wtf/NoVirtualDestructorBase.h: Added.
* Source/WTF/wtf/text/AtomStringImpl.h:
* Source/WTF/wtf/text/ExternalStringImpl.h:
* Source/WTF/wtf/text/StringImpl.h:
* Source/WTF/wtf/text/SymbolImpl.h:
* Source/WTF/wtf/text/UniquedStringImpl.h:
* Source/WebCore/SaferCPPExpectations/RefCntblBaseVirtualDtorExpectations:
* Source/WebCore/css/CSSValue.h:
* Source/WebCore/dom/NodeRareData.h:

Canonical link: <a href="https://commits.webkit.org/293307@main">https://commits.webkit.org/293307@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4245ea09cfc6c81686735259f39e58148ae30ce4

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/98483 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/18116 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/8348 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/103604 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/49015 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/100527 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/18409 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/26567 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/74971 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/32128 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/101487 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/13964 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/88950 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/55330 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/13744 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/6914 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/48453 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/91170 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/83700 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/6992 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/105977 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/97112 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/25573 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/18620 "Found 1 new test failure: imported/w3c/web-platform-tests/html/browsers/the-window-object/open-close/open-features-tokenization-noreferrer.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/83943 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/25949 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/85150 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/83427 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/28072 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/5742 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/19229 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/15966 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/25531 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/30712 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/120729 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/25349 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/33788 "Found 19737 jsc stress test failures: ChakraCore.yaml/ChakraCore/test/Array/LastUsedSegmentHasNULLElement.js.default, ChakraCore.yaml/ChakraCore/test/Array/array_ctr.js.default, ChakraCore.yaml/ChakraCore/test/Array/array_fastinit.js.default, ChakraCore.yaml/ChakraCore/test/Array/array_includes.js.default, ChakraCore.yaml/ChakraCore/test/Array/array_indexOf.js.default ..., JSC test binary failure: testapi") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/28669 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/26924 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WTF/wtf/NoVirtualDestructorBase.h:
```diff
@@ -0,0 +1,29 @@
+
+#pragma once
+
+class NoVirtualDestructorBase {
```
I think this class definitely needs a comment to explain what it's used for. It's weird enough of a use case :)

rniwa replied:
ok, added.

cdumez reviewed (commented):
I thought CSSValue was one of the big ones but it's not addressed in this PR?

cdumez commented on Source/WTF/wtf/text/StringImpl.h:
```diff
@@ -145,7 +146,7 @@ struct StringStats {
 #endif
 
-class STRING_IMPL_ALIGNMENT StringImplShape {
+class STRING_IMPL_ALIGNMENT StringImplShape : public NoVirtualDestructorBase {
```
Does the subclassing really have to be `public`?

rniwa replied:
Yes. In theory, we could support "private" inheritance as well but I didn't do that. I also think public inheritance makes more sense because this communicates the protocol this and all its subclasses conform to.

cdumez replied:
I have concerns given that we wouldn't want someone casting a `StringImplShape` to a `NoVirtualDestructorBase` type, which would lead to destruction issues.

rniwa replied:
Why would anyone cast to it?

cdumez replied:
To some this might look like a new convenient common base class for previously unrelated types. The cast can be implicit by just assigning to a `NoVirtualDestructorBase*`.
```
HashSet<std::unique_ptr<NoVirtualDestructorBase>> objects;
objects.add(WTFMove(aStringImplShare));
objects.add(WTFMove(aJSONValue));
objects.add(WTFMove(anotherNoVirtualDestructorBaseSubclass));
```

I hope people wouldn't try but you never know. It seemed unfortunate to me to leave the door opened to it happening. Having a static analysis workaround open the door to doing new things in C++ that aren't safe is not ideal.

That said, if others don't think this is an issue, go for it. I will not die on this hill. I simply would have never designed it this way myself.

At the very least, I do think we should be the `NoVirtualDestructorBase` constructor and destructor protected instead of public.

rniwa replied:
sure, we can make constructors & destructors protected.

cdumez commented on Source/WTF/wtf/JSONValues.h:
```diff
@@ -52,7 +53,7 @@ class Object;
 
 // FIXME: unify this JSON parser with JSONParse in JavaScriptCore.
-class WTF_EXPORT_PRIVATE Value : public RefCounted<Value> {
+class WTF_EXPORT_PRIVATE Value : public RefCounted<Value>, public NoVirtualDestructorBase {
```
For what it's worth, I don't like the `NoVirtualDestructorBase` name but I assume it's hardcoded in the static analysis? Ideally the name would be something like "SuppressVirtualDestructorChecks" or "SuppressVirtualDestructorStaticAnalysis". But at least we have a comment next to `NoVirtualDestructorBase` explaining that it is about static analysis..

rniwa replied:
yeah, it's hard-coded.

rniwa committed:
Introduce NoVirtualDestructorBase and use it in various WTF classes
https://bugs.webkit.org/show_bug.cgi?id=290651

Reviewed by Chris Dumez.

Introduce NoVirtualDestructorBase and replace the use of
SUPPRESS_REFCOUNTED_WITHOUT_VIRTUAL_DESTRUCTOR with it in various WTF classes.

Suppress the unsafe cast warning in SymbolImpl::StaticSymbolImpl now that
this class is checked by the static analyzer.

* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/CMakeLists.txt:
* Source/WTF/wtf/Compiler.h:
* Source/WTF/wtf/JSONValues.h:
* Source/WTF/wtf/NoVirtualDestructorBase.h: Added.
* Source/WTF/wtf/text/AtomStringImpl.h:
* Source/WTF/wtf/text/ExternalStringImpl.h:
* Source/WTF/wtf/text/StringImpl.h:
* Source/WTF/wtf/text/SymbolImpl.h:
* Source/WTF/wtf/text/UniquedStringImpl.h:
* Source/WebCore/SaferCPPExpectations/RefCntblBaseVirtualDtorExpectations:
* Source/WebCore/css/CSSValue.h:
* Source/WebCore/dom/NodeRareData.h:

Canonical link: https://commits.webkit.org/293307@main
